### ISSUE TITLE
Failed to create cluster by self-builded kind-node image

### ISSUE THREAD

**Issue by cmssczy** [kubernetes-sigs/kind#2493]
I can create a cluster successfully when using offical image(eg. kindest/node:v1.21.1). But It will be failed to create a cluster with the image which was build by myself, and the panic log is as follow. 

**What happened**:
```bash
kind create cluster --retain  --config new-kind.yaml 
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:latest) 🖼 
 ✓ Preparing nodes 📦 📦  
 ✓ Writing configuration 📜 
 ✗ Starting control-plane 🕹️ 
ERROR: failed to create cluster: failed to init node with kubeadm: command "docker exec --privileged kind-control-plane kubeadm init --skip-phases=preflight --config=/kind/kubeadm.conf --skip-token-print --v=6" failed with error: exit status 1
Command Output: I1012 07:15:28.620246     182 initconfiguration.go:248] loading configuration from "/kind/kubeadm.conf"
[config] WARNING: Ignored YAML document with GroupVersionKind kubeadm.k8s.io/v1beta2, Kind=JoinConfiguration
I1012 07:15:28.779378     182 common.go:117] WARNING: tolerating control plane version v1.20.0-alpha.0.11100+802c8dfd69c550-dirty, assuming that k8s version 1.19.0 is not released yet
[init] Using Kubernetes version: v1.20.0-alpha.0.11100+802c8dfd69c550-dirty
[certs] Using certificateDir folder "/etc/kubernetes/pki"
I1012 07:15:28.781093     182 certs.go:111] creating a new certificate authority for ca
[certs] Generating "ca" certificate and key
I1012 07:15:29.190809     182 certs.go:487] validating certificate period for ca certificate
[certs] Generating "apiserver" certificate and key
[certs] apiserver serving cert is signed for DNS names [kind-control-plane kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local localhost] and IPs [10.96.0.1 172.19.0.3 127.0.0.1]
[certs] Generating "apiserver-kubelet-client" certificate and key
I1012 07:15:30.011998     182 certs.go:111] creating a new certificate authority for front-proxy-ca
[certs] Generating "front-proxy-ca" certificate and key
I1012 07:15:30.460770     182 certs.go:487] validating certificate period for front-proxy-ca certificate
[certs] Generating "front-proxy-client" certificate and key
I1012 07:15:31.132299     182 certs.go:111] creating a new certificate authority for etcd-ca
[certs] Generating "etcd/ca" certificate and key
I1012 07:15:31.264557     182 certs.go:487] validating certificate period for etcd/ca certificate
[certs] Generating "etcd/server" certificate and key
[certs] etcd/server serving cert is signed for DNS names [kind-control-plane localhost] and IPs [172.19.0.3 127.0.0.1 ::1]
[certs] Generating "etcd/peer" certificate and key
[certs] etcd/peer serving cert is signed for DNS names [kind-control-plane localhost] and IPs [172.19.0.3 127.0.0.1 ::1]
[certs] Generating "etcd/healthcheck-client" certificate and key
[certs] Generating "apiserver-etcd-client" certificate and key
I1012 07:15:33.378561     182 certs.go:77] creating new public/private key files for signing service account users
[certs] Generating "sa" key and public key
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
I1012 07:15:34.359068     182 kubeconfig.go:103] creating kubeconfig file for admin.conf
[kubeconfig] Writing "admin.conf" kubeconfig file
I1012 07:15:35.256869     182 kubeconfig.go:103] creating kubeconfig file for kubelet.conf
[kubeconfig] Writing "kubelet.conf" kubeconfig file
I1012 07:15:37.407592     182 kubeconfig.go:103] creating kubeconfig file for controller-manager.conf
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
I1012 07:15:37.830218     182 kubeconfig.go:103] creating kubeconfig file for scheduler.conf
[kubeconfig] Writing "scheduler.conf" kubeconfig file
I1012 07:15:39.537276     182 kubelet.go:65] Stopping the kubelet
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Starting the kubelet
[control-plane] Using manifest folder "/etc/kubernetes/manifests"
[control-plane] Creating static Pod manifest for "kube-apiserver"
I1012 07:15:40.223298     182 manifests.go:99] [control-plane] getting StaticPodSpecs
I1012 07:15:40.239912     182 certs.go:487] validating certificate period for CA certificate
I1012 07:15:40.241203     182 manifests.go:125] [control-plane] adding volume "ca-certs" for component "kube-apiserver"
I1012 07:15:40.242170     182 manifests.go:125] [control-plane] adding volume "etc-ca-certificates" for component "kube-apiserver"
I1012 07:15:40.243235     182 manifests.go:125] [control-plane] adding volume "k8s-certs" for component "kube-apiserver"
I1012 07:15:40.244716     182 manifests.go:125] [control-plane] adding volume "usr-local-share-ca-certificates" for component "kube-apiserver"
I1012 07:15:40.244796     182 manifests.go:125] [control-plane] adding volume "usr-share-ca-certificates" for component "kube-apiserver"
I1012 07:15:40.346201     182 manifests.go:154] [control-plane] wrote static Pod manifest for component "kube-apiserver" to "/etc/kubernetes/manifests/kube-apiserver.yaml"
[control-plane] Creating static Pod manifest for "kube-controller-manager"
I1012 07:15:40.365630     182 manifests.go:99] [control-plane] getting StaticPodSpecs
I1012 07:15:40.366645     182 manifests.go:125] [control-plane] adding volume "ca-certs" for component "kube-controller-manager"
I1012 07:15:40.366747     182 manifests.go:125] [control-plane] adding volume "etc-ca-certificates" for component "kube-controller-manager"
I1012 07:15:40.366800     182 manifests.go:125] [control-plane] adding volume "flexvolume-dir" for component "kube-controller-manager"
I1012 07:15:40.366848     182 manifests.go:125] [control-plane] adding volume "k8s-certs" for component "kube-controller-manager"
I1012 07:15:40.366896     182 manifests.go:125] [control-plane] adding volume "kubeconfig" for component "kube-controller-manager"
I1012 07:15:40.366943     182 manifests.go:125] [control-plane] adding volume "usr-local-share-ca-certificates" for component "kube-controller-manager"
I1012 07:15:40.366991     182 manifests.go:125] [control-plane] adding volume "usr-share-ca-certificates" for component "kube-controller-manager"
I1012 07:15:40.367757     182 manifests.go:154] [control-plane] wrote static Pod manifest for component "kube-controller-manager" to "/etc/kubernetes/manifests/kube-controller-manager.yaml"
[control-plane] Creating static Pod manifest for "kube-scheduler"
I1012 07:15:40.374828     182 manifests.go:99] [control-plane] getting StaticPodSpecs
I1012 07:15:40.376605     182 manifests.go:125] [control-plane] adding volume "kubeconfig" for component "kube-scheduler"
I1012 07:15:40.377730     182 manifests.go:154] [control-plane] wrote static Pod manifest for component "kube-scheduler" to "/etc/kubernetes/manifests/kube-scheduler.yaml"
[etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
I1012 07:15:40.382440     182 local.go:65] [etcd] wrote Static Pod manifest for a local etcd member to "/etc/kubernetes/manifests/etcd.yaml"
I1012 07:15:40.398756     182 waitcontrolplane.go:91] [wait-control-plane] Waiting for the API server to be healthy
I1012 07:15:40.400953     182 loader.go:372] Config loaded from file:  /etc/kubernetes/admin.conf
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
I1012 07:15:40.492118     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 4 milliseconds
I1012 07:15:41.004278     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:41.498622     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:41.997366     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:42.505348     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:43.001522     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 5 milliseconds
I1012 07:15:43.499258     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:43.994750     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:44.494230     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:15:44.994179     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:19:39.994099     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:19:40.494124     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I1012 07:19:40.494524     182 round_trippers.go:460] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds

	Unfortunately, an error has occurred:
		timed out waiting for the condition

	This error is likely caused by:
		- The kubelet is not running
		- The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)

	If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
		- 'systemctl status kubelet'
		- 'journalctl -xeu kubelet'

	Additionally, a control plane component may have crashed or exited when started by the container runtime.
	To troubleshoot, list all containers using your preferred container runtimes CLI.

	Here is one example how you may list all Kubernetes containers running in cri-o/containerd using crictl:
		- 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps -a | grep kube | grep -v pause'
		Once you have found the failing container, you can inspect its logs with:
		- 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock logs CONTAINERID'

couldn't initialize a Kubernetes cluster
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runWaitControlPlanePhase
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go:118
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:856
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:974
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:902
k8s.io/kubernetes/cmd/kubeadm/app.Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
	_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
	/usr/local/go/src/runtime/proc.go:255
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1581
error execution phase wait-control-plane
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:235
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:856
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:974
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:902
k8s.io/kubernetes/cmd/kubeadm/app.Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
	_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
	/usr/local/go/src/runtime/proc.go:255
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1581

```
**kind export logs**:

[kind.zip](https://github.com/kubernetes-sigs/kind/files/7327801/kind.zip)

**What you expected to happen**:
I expected a cluster creating and running

**How to reproduce it (as minimally and precisely as possible)**:

- build my own image
```bash
# kind build node-image
Starting to build Kubernetes
+++ [1012 02:20:21] Verifying Prerequisites....
+++ [1012 02:20:23] Building Docker image kube-build:build-22f27c1af5-5-v1.23.0-go1.17.1-bullseye.1
+++ [1012 02:20:30] Creating data container kube-build-data-22f27c1af5-5-v1.23.0-go1.17.1-bullseye.1
+++ [1012 02:20:55] Syncing sources to container
+++ [1012 02:21:27] Running build command...
+++ [1012 02:21:48] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
Generating prerelease lifecycle code for 28 targets
+++ [1012 02:22:32] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
Generating deepcopy code for 237 targets
+++ [1012 02:22:49] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
Generating defaulter code for 94 targets
+++ [1012 02:23:10] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
Generating conversion code for 130 targets
+++ [1012 02:23:49] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
Generating openapi code for CODEGEN
Generating openapi code for SAMPLEAPISERVER
+++ [1012 02:24:30] Building go targets for linux/amd64:
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kube-scheduler
    cmd/kube-proxy
    cmd/kubeadm
    cmd/kubectl
    cmd/kubelet
+++ [1012 02:51:15] Syncing out of container
+++ [1012 02:51:37] Building images: linux-amd64
+++ [1012 02:51:38] Starting docker build for image: kube-apiserver-amd64
+++ [1012 02:51:38] Starting docker build for image: kube-controller-manager-amd64
+++ [1012 02:51:38] Starting docker build for image: kube-scheduler-amd64
+++ [1012 02:51:38] Starting docker build for image: kube-proxy-amd64
+++ [1012 02:52:16] Deleting docker image k8s.gcr.io/kube-scheduler-amd64:v1.20.0-alpha.0.11100_802c8dfd69c550-dirty
+++ [1012 02:52:29] Deleting docker image k8s.gcr.io/kube-proxy-amd64:v1.20.0-alpha.0.11100_802c8dfd69c550-dirty
+++ [1012 02:52:31] Deleting docker image k8s.gcr.io/kube-controller-manager-amd64:v1.20.0-alpha.0.11100_802c8dfd69c550-dirty
+++ [1012 02:52:38] Deleting docker image k8s.gcr.io/kube-apiserver-amd64:v1.20.0-alpha.0.11100_802c8dfd69c550-dirty
+++ [1012 02:52:38] Docker builds done
+++ [1012 02:52:38] Verifying Prerequisites....
+++ [1012 02:52:44] Building Docker image kube-build:build-22f27c1af5-5-v1.23.0-go1.17.1-bullseye.1
+++ [1012 02:52:50] Syncing sources to container
+++ [1012 02:52:55] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [1012 02:52:55] Running build command...
Generating prerelease lifecycle code for 28 targets
Generating deepcopy code for 237 targets
Generating defaulter code for 94 targets
Generating conversion code for 130 targets
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
Generating openapi code for CODEGEN
Generating openapi code for SAMPLEAPISERVER
+++ [1012 02:55:08] Building go targets for linux/amd64:
    cmd/kubeadm
    cmd/kubectl
    cmd/kubelet
+++ [1012 02:56:24] Syncing out of container
Finished building Kubernetes
Building node image ...
Building in kind-build-1634021843-1975416647
sha256:ad9e954da9f87e3705aefa97af72e8700020ea24b685ae4a58cfa1f53418714b
Image build completed.

```

**Anything else we need to know?**:
It will be successful when I create cluster using official image.
```bash
 kind create cluster --config new-kind.yaml 
Creating cluster "newkind" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼 
 ✓ Preparing nodes 📦 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
 ✓ Joining worker nodes 🚜 
Set kubectl context to "kind-newkind"
You can now use your cluster with:

kubectl cluster-info --context kind-newkind

Thanks for using kind! 😊

```
**Environment:**

- kind version: (use `kind version`):
`kind v0.11.1 go1.17.2 linux/amd64`
- Kubernetes version: (use `kubectl version`):
- Docker version: (use `docker info`):
  it can be seen in exported-logs.
- OS (e.g. from `/etc/os-release`):
I create a virtual machine by VirtualBox and installed Ubuntu.
```bash
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

**Comment by cmssczy** [comment]
actually I have pulled the `pause` image before I run command `kind create cluster`. @neolit123 
```bash
docker images
k8s.gcr.io/pause   3.5   ed210e3e4a5b   6 months ago   683kB
```

**Comment by neolit123** [comment]
> actually I have pulled the pause image before I run command kind create cluster

it must be in the kind node image, not on the host.

**Comment by cmssczy** [comment]
> > actually I have pulled the pause image before I run command kind create cluster
> 
> it must be in the kind node image, not on the host.

Got it. Thanks.
I have set the proxy for docker and my local host. But how to set the network proxy for `kind`? 
I have to do this because I'm working in China and it will block some URL domain names for reasons that are well known. 😂

**Comment by cmssczy** [comment]
Why using the official `kind-node` image will work? It doesn't need to pull `pause` image?

**Comment by BenTheElder** [comment]
> Did you mean that kind build node-image should use the HTTP*_PROXY which was set by docker when ctr pull the image? @BenTheElder

I didn't catch anything about it being set by docker 🤔 , but I would expect `kind build node-image` to respect HTTP_PROXY set in the environment `kind` is called in by passing it through, except I don't think it does currently.

> I didn't pass the HTPP*_PROXY to kind build node-image at the very begining. So in the kind-control-plane container we could not see the pause image by ctr -n=k8s.io image list.

HTTP*_PROXY should be set if there's a proxy being used, building is also a networked task, and this seems to be the only reasonable "standard" for passing proxy info to these tools.

I think if you *had* done that, it still wouldn't have pulled the image successfully because we don't pass down that env to the containerd instance pulling the image inside the build environment.

> After I set the HTTP*_PROXY by export HTTPS_PROXY="http://ip:port" HTTP_PROXY="http://ip:port" and then run kind build node-image, it will be success.

What's not clear is can you:
1. `kind build node-image` with HTTP*_PROXY set
2. `kind create cluster` with that image *without* HTTP*_PROXY set, or even network access.

2) only works if 1) actually respects the settings fully. 
For example you can see that even if you disable the internet on your machine, if you run `kind create cluster` (no options) it will succeed as long as the core `kindest/node` image is already downloaded, it's a single artifact for offline.

It sounds like what actually happened is you set HTTP*_PROXY in a shell, then did build + create cluster, and actually the HTTP_PROXY allowed it to pull the image during create that was not bundled properly during build still.

If so: That's still incorrect behavior, leading to slower startup time and breaking the ability to run the custom kindest/node image offline.

**Comment by cmssczy** [comment]
> * After I set the `HTTP*_PROXY` by `export HTTPS_PROXY="http://ip:port" HTTP_PROXY="http://ip:port"` and then run `kind build node-image`, it will be success.

Sorry, here is a mistake, it should be `run kind create cluster --config kind.yaml` not`run kind build node-image`. Maybe this mistake will cause misunderstanding. Sorry for that.
I think there is no need to set the proxy when build the image. This problem will only arise when create kind cluster. 

Here is the detailed progress:
1. Before I started to build image and create `kind` cluster, I set the network proxy in my Ubuntu system setting.
2. Then I tried to build image by `kind build node-image` and it successed.
3. Then I tried to create a cluster by `kind create cluster --config kind.yaml` but it failed because it could not pull the `pause` image
4.  I set the HTTP*_PROXY by export HTTPS_PROXY="http://ip:port" HTTP_PROXY="http://ip:port" and then create the cluster, it successed.

@BenTheElder

**Comment by BenTheElder** [comment]
Yes, that's what I thought.

It *should* work if you do:
1. set your network proxy
2. set temporarily HTTP*_PROXY and build an image `HTTP_PROXY=... HTTPS_PROXY=... kind build node-image`
3. *without* HTTP*_PROXY set (or even with the internet disconnected) `kind create cluster --image=kindest/node:latest` (create a cluster with the image you just built) should just work.

But right now I don't think 2) works correctly. It should be fixed.

We can't avoid needing to set `HTTP*_PROXY` at all. But your current workaround means during cluster creation some of the images that should have been pulled at node-image build time are being pulled during cluster creation, which is not ideal:
- it means if you share / use this image it won't work offline, unlike one of our official images (that are built w/o proxy issues)
- it means that the startup time is negatively impacted for each cluster, because the images are not pre-populated

We should ensure this workflow works.

**Comment by cmssczy** [comment]
> But your current workaround means during cluster creation some of the images that should have been pulled at node-image build time are being pulled during cluster creation, which is not ideal:

Yes, that's the point.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/issues/2562 will cover the remainder of this going forward, thanks!

**Comment by yy158775** [comment]
But now how to pull the pause image in internal network when create custom cluster image.If I set proxy when `kind create cluster`,it will lead to slower startup time. @BenTheElder

**Comment by BenTheElder** [comment]
For building images: https://github.com/kubernetes-sigs/kind/issues/2562

Once built node images shouldn't need to pull pause, that's a bug if it happens.

**Comment by yy158775** [comment]
Can I put images into built node images to avoid pulling images during cluster creation?

**Comment by BenTheElder** [comment]
> Can I put images into built node images to avoid pulling images during cluster creation?

This is how kind works normally. All of the required images are pre-loaded. In this issue we discovered that when building in a proxied environment that's broken in that it will continue on and fail to preload. #2562 tracks ensuring the proxy is passed to image pulling at build time, and we fixed that it would silently fail already.

At runtime if you have additional images, see `kind load` to import images into the node without pulling to the node.
